snakemake/deployment/docker.py

```diff
@@ -0,0 +1,119 @@
+__author__ = "Johannes Köster"
+__copyright__ = "Copyright 2015-2020, Johannes Köster"
+__email__ = "johannes.koester@uni-due.de"
+__license__ = "MIT"
+
+import os
+import re
+import requests
+
+
+class DockerContainer:
+    """A Docker Container is a thin wrapper to a Docker Registry API to
+       primarily determine if a container exists to be used for a workflow
+    """
+
+    def __init__(self, name):
+        self.name = name
+        self.uri = None
+        if self.exists():
+            self.uri = "docker://%s" % self.name
+
+    @property
+    def version(self):
+        """Derive a container verison. Will by default be latest (the tag)
+           but can be a different tag or full hash.
+        """
+        # Does the container have a version (tag or hash?)
+        version = "latest"
+        if "@" in self.name:
+            container_name, version = self.name.split("@", 1)
+        elif ":" in self.name:
+            container_name, version = self.name.split(":", 1)
+        return version
+
+    @property
+    def registry(self):
+        """Return the container registry. We default to Docker Hub unless
+           another (quay.io) is defined
+        """
+        registry = "https://index.docker.io/v2"
+        if "quay.io" in self.name:
+            registry = "https://quay.io/api/v1"
+        return registry
+
+    @property
+    def repository(self):
+        """The repository is the username/reponame combination, without
+           the registry. E.g., quay.io/name/repo --> name/repo
+        """
+        container_name = self.name.replace("quay.io/", "", 1)
+        return re.split("(:|@)", container_name, 1)[0]
+
+    def exists(self):
+        """Return True if the container exists under CONTAINER_PREFIX 
+           (quay.io/snakemake-wrappers) otherwise return False
+        """
+        # Request for the manifest
+        # GET /v2/<name>/manifests/<reference>
+        base = "%s/%s/manifests/%s" % (self.registry, self.repository, self.version)
+        headers = {"Docker-Distribution-API-Version": "registry/2.0"}
+        response = requests.get(base, headers=headers)
+
+        # Repository does not exist
+        if response.status_code == 404:
```

Comment:
I tested the 404 returned for Quay.io, and if a container doesn't exist it will return 404. I still need to test when a container does exist (if Quay also requires an authentication token). For Docker Hub, if the container doesn't exist _or_ it doesn't have a token, it returns 401 either way.